ports/esp32/machine_pwm.c

```diff
@@ -244,53 +287,53 @@ static uint32_t ledc_find_suitable_duty_resolution(uint32_t src_clk_freq, uint32
 }
 #endif
 
-static void set_freq(machine_pwm_obj_t *self, unsigned int freq, ledc_timer_config_t *timer) {
-    esp_err_t err;
+static void set_freq(machine_pwm_obj_t *self, unsigned int freq, ledc_timer_config_t *timer, ledc_clk_cfg_t led_src_clock) {
+
     if (freq != timer->freq_hz) {
-        // Configure the new frequency and resolution
-        timer->freq_hz = freq;
 
-        #if SOC_LEDC_SUPPORT_PLL_DIV_CLOCK
-        timer->clk_cfg = LEDC_USE_PLL_DIV_CLK;
-        #elif SOC_LEDC_SUPPORT_APB_CLOCK
-        timer->clk_cfg = LEDC_USE_APB_CLK;
-        #elif SOC_LEDC_SUPPORT_XTAL_CLOCK
-        timer->clk_cfg = LEDC_USE_XTAL_CLK;
-        #else
-        #error No supported PWM / LEDC clocks.
-        #endif
-        #if SOC_LEDC_SUPPORT_REF_TICK
-        if (freq < EMPIRIC_FREQ) {
-            timer->clk_cfg = LEDC_USE_REF_TICK;
-        }
-        #endif
-        uint32_t src_clk_freq = 0;
+        uint32_t src_clock_freq;
+
         #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 1, 0)
-        err = esp_clk_tree_src_get_freq_hz(timer->clk_cfg, ESP_CLK_TREE_SRC_FREQ_PRECISION_CACHED, &src_clk_freq);
-        if (err != ESP_OK) {
-            mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("unable to query source clock frequency %d"), (int)timer->clk_cfg);
+
+        if (esp_clk_tree_src_get_freq_hz(led_src_clock, ESP_CLK_TREE_SRC_FREQ_PRECISION_APPROX, &src_clock_freq) != ESP_OK) {
+            mp_raise_ValueError(MP_ERROR_TEXT("Error on getting reference clock frequency (FREQ_PRECISION_APPROX)."));
         }
+
+        uint32_t res = ledc_find_suitable_duty_resolution(src_clock_freq, freq);
         #else
-        // Simplified fallback logic for IDF V5.0.x, for targets with APB only.
-        src_clk_freq = APB_CLK_FREQ; // 80 MHz
-        #if SOC_LEDC_SUPPORT_REF_TICK
-        if (timer->clk_cfg == LEDC_USE_REF_TICK) {
-            src_clk_freq = REF_CLK_FREQ; // 1 MHz
+
+        if (led_src_clock == LEDC_USE_APB_CLK) {
+            src_clock_freq = APB_CLK_FREQ;
+        } else if (led_src_clock == LEDC_USE_RC_FAST_CLK) {
+            src_clock_freq = SOC_CLK_RC_FAST_FREQ_APPROX;
+        } else if (led_src_clock == LEDC_USE_REF_TICK) {
+            src_clock_freq = REF_CLK_FREQ;
         }
-        #endif // SOC_LEDC_SUPPORT_REF_TICK
-        #endif // ESP_IDF_VERSION
 
-        timer->duty_resolution = ledc_find_suitable_duty_resolution(src_clk_freq, timer->freq_hz);
+        uint32_t res = ledc_find_suitable_duty_resolution(src_clock_freq, freq);
+
+        #endif
+
+        timer->duty_resolution = res;
```

Comment:
It may be replaced by
```
timer->duty_resolution = ledc_find_suitable_duty_resolution(src_clock_freq, freq);
```
and remove two upper
`uint32_t res = ledc_find_suitable_duty_resolution(src_clock_freq, freq);`

Reply:
Yes I will do :-)

Reply:
Done :-)